src/openai/lib/azure.py

```diff
@@ -0,0 +1,210 @@
+
+        return super()._build_request(options)
+
+
+class AzureOpenAI(BaseAzureClient, OpenAI):
+    @overload
+    def __init__(
+        self,
+        *,
+        api_version: str,
+        endpoint: str,
+        deployment: str | None = None,
+        api_key: str | None = None,
+        organization: str | None = None,
+        timeout: float | Timeout | None | NotGiven = NOT_GIVEN,
+        max_retries: int = DEFAULT_MAX_RETRIES,
+        default_headers: Mapping[str, str] | None = None,
+        default_query: Mapping[str, object] | None = None,
+        http_client: httpx.Client | None = None,
+        _strict_response_validation: bool = False,
+    ) -> None:
+        ...
+
+    @overload
+    def __init__(
+        self,
+        *,
+        api_version: str,
+        base_url: str,
+        api_key: str | None = None,
+        organization: str | None = None,
+        timeout: float | Timeout | None | NotGiven = NOT_GIVEN,
+        max_retries: int = DEFAULT_MAX_RETRIES,
+        default_headers: Mapping[str, str] | None = None,
+        default_query: Mapping[str, object] | None = None,
+        http_client: httpx.Client | None = None,
+        _strict_response_validation: bool = False,
+    ) -> None:
+        ...
+
+    def __init__(
+        self,
+        *,
+        api_version: str,
+        endpoint: str | None = None,
+        deployment: str | None = None,
+        api_key: str | None = None,
+        organization: str | None = None,
+        base_url: str | None = None,
+        timeout: float | Timeout | None | NotGiven = NOT_GIVEN,
+        max_retries: int = DEFAULT_MAX_RETRIES,
+        default_headers: Mapping[str, str] | None = None,
+        default_query: Mapping[str, object] | None = None,
+        http_client: httpx.Client | None = None,
+        _strict_response_validation: bool = False,
+    ) -> None:
+        if api_key is None:
+            api_key = os.environ.get("AZURE_OPENAI_API_KEY")
+        if api_key is None:
+            raise OpenAIError(
+                "The api_key client option must be set either by passing api_key to the client or by setting the AZURE_OPENAI_API_KEY environment variable"
+            )
+
+        if default_headers is None:
+            default_headers = {"api-key": api_key}
+        else:
+            default_headers = {"api-key": api_key, **default_headers}
+
+        if default_query is None:
+            default_query = {"api-version": api_version}
+        else:
+            default_query = {"api-version": api_version, **default_query}
+
+        if base_url is None:
+            if endpoint is None:
+                raise ValueError("If base_url is not given, then endpoint must be given")
+
+            if deployment is not None:
+                base_url = f"https://{endpoint}.openai.azure.com/openai/deployments/{deployment}"
+            else:
+                base_url = f"https://{endpoint}.openai.azure.com/openai"
```

Comment:
The endpoint passed will look something like "https://my-resource.openai.azure.com" so we don't need to hardcode the extra bits.

```suggestion
            if deployment is not None:
                base_url = f"{endpoint}/openai/deployments/{deployment}"
            else:
                base_url = f"{endpoint}/openai"
```

Reply:
Ah okay, in that case is there much value in providing the separate `endpoint` argument then? It would likely be simpler to require users to just pass `base_url` instead, e.g.

```py
AzureOpenAI(
  api_version='...',
  base_url='https://my-resource.openai.azure.com/openai'
)
```

Reply:
The main value in providing the base_url and (endpoint + deployment) as separate overloads is that all existing documentation/portal etc. is exposing the endpoint without the openai segment.

Reply:
@johanste After some discussion internally we think it would be a much better user experience if we provided an argument that was just the resource name. Especially in the CLI, e.g.
```shell
openai --azure-endpoint="https://my-resource.openai.azure.com/openai" --azure-version="2023-07-01-preview" api chat.completions.create ...
```

compared to
```shell
openai --azure-resource="my-resource" --azure-version="2023-07-01-preview" api chat.completions.create ...
```

But we're not opposed to also providing an `endpoint` argument if you think that would be beneficial.

Reply:
Endpoint argument is (or I assume very likely will be) a requirement. Azure supports multiple separate cloud regions with different dns names.

Reply:
And I think we can start with endpoint only. It is easier to listen to customer feedback and see if they want something more than that than it is taking it away if we ever shipped it.

Reply:
We now have support for both `resource` and `endpoint`; I'd be inclined to continue to support both. All the Azure docs make it look like there's only one "variable" within `endpoint`, which is `resource`. 

> Azure supports multiple separate cloud regions with different dns names.

Can you share examples of what these look like?

Reply:
Sorry to interject here, but just want to +1 only having support for `endpoint`. This is what users copy/paste from the Portal, CLI, etc. @johanste mentioned the issue with different cloud regions, but also want to point out that we need to support multi-service AI resources as well (which includes Azure OpenAI + other Azure AI services in the same resource). Multi-service won't look like "https://my-resource.openai.azure.com/" and we won't be able to discern if it is a multi-service resource based only on the passed resource name. I could see this leading to some confusing errors for users...

Reply:
No apologies, thanks for weighing in @kristapratico ! We're less familiar with Azure and mostly know what we've seen from the docs, which always make it seem to me as a user that the resource is the only thing that changes. So appreciate your input here.

What do multi-service and multi-region URLs look like?

Reply:
My sense is that if a user is on a different service/region, they will get a 404 error if they only pass `resource`, and it should be fairly clear once they look at the URLs at hand that they need to pass `endpoint` or `base_url` instead (both of which are available and easy to find)

Reply:
Examples on difference in endpoints across global azure and azure government clouds: https://learn.microsoft.com/en-us/azure/azure-government/compare-azure-government-global-azure.

I don't have numbers of percentage of users, but we generally expect users to pass in the endpoint. In theory, you could have the same resource name in two different cloud instances which would lead to differently confusing error messages.

Reply:
For multi-service, I've seen "https://{region}.api.cognitive.microsoft.com/" and "https://{my-resource}.cognitiveservices.azure.com/". Both are valid. For multicloud, I defer to @johanste, but I believe it would be something like: "https://{my-resource}.openai.azure.us/" or "https://{my-resource}.openai.azure.cn/". I'm not sure if I'm covering all possibilities here, though.

Not sure on percentage of users on multi-service or multi-cloud. I can see if I can get that data. Definitely agree on design decisions driven by customer data/feedback. When it comes to convenience things, I tend to fall under the  "wait for customer feedback" group, since if we do get signal that it accidentally leads users down the 404 path, we won't be able to remove it so easily.

Please excuse me for making the "Zen of Python" argument 😆, however, I think having `base_url`, `endpoint`, and `resource` as 3 options paints a non-obvious way of how to do one thing. `base_url` and `endpoint` seem sufficient, IMO.

Reply:
Thanks for all the context! We're happy with removing it, I'll update the PR shortly!

Reply:
Agreed, thank you both!